examples/server/utils.hpp

```diff
@@ -167,6 +167,34 @@ static T json_value(const json &body, const std::string &key, const T &default_v
         : default_value;
 }
 
+inline std::string format_llama2(std::vector<json> messages)
+{
+    std::ostringstream output;
+    bool is_inside_turn = false;
+
+    for (auto it = messages.begin(); it != messages.end(); ++it) {
+        if (!is_inside_turn) {
+            output << "[INST] ";
+        }
+        std::string role    = json_value(*it, "role", std::string("user"));
+        std::string content = json_value(*it, "content", std::string(""));
+        if (role == "system") {
+            output << "<<SYS>>\n" << content << "\n<<SYS>>\n\n";
+            is_inside_turn = true;
+        } else if (role == "user") {
+            output << content << " [/INST]";
+            is_inside_turn = true;
+        } else {
+            output << " " << content << " </s>";
```

Comment:
I don't think there should be a space in front of the eos token.

Reply:
As discussed in https://github.com/ggerganov/llama.cpp/pull/5425#discussion_r1484880949 , the space is not consistent among different variations of this template. I leave it here to be sure. It changes nothing on the behavior of the model.

Ideally if I had access to `tokenizer.chat_template`, I can adapt the template the template to exactly what the model need (i.e. spaces, `<<SYS>>`,...). But since I don't have access to that part for now, I kinda have to make a "catch-all" template.

And also, what is important in this template is the placement of `[INST]`, `[/INST]` and `</s>`. A model that never need trained to understand chatml will not understand that the message should be encapsulated between `<|im_start|>` and `<|im_end|>`. It will kind of "repeat what's already in the conversation".

As long as the model see `[INST] some user prompt [/INST] some assistant responses </s>`, it will behave just fine. The spaces does not matter, that's what I experience while testing this PR.